### PR TITLE
eni: fix new node not immediately triggers the creation of Eni before instances api  sync

### LINKED ISSUE #20678: ENI: new node not does not immediately trigger the creation of Eni
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

1. New node add to cluster, it not immediately trigger the creation of Eni.  Until resync period is triggered, eni be created in new node. so agent in waiting eni create max duration is 1min. 
2. It happend in aws and alibabacloud and azure.
3. it also find log in issue https://github.com/cilium/cilium/issues/18239#issuecomment-992927403 https://github.com/cilium/cilium/issues/14028 https://github.com/cilium/cilium/issues/11560#issuecomment-902693365
```
Instance not found! Please delete corresponding ciliumnode if instance has already been deleted
``` 
4.  `instance does not exist` in azure
5. Because it not update `instanceManger.instances` when new node find. So in `ops.ResyncInterfacesAndIPs` return error in aws and alibabacloud.
https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/aws/eni/node.go#L509-L542
https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/aws/eni/instances.go#L216-L230
https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/ipam/types/types.go#L433-L449

### Cilium Version

master https://github.com/cilium/cilium/tree/94ea7fc93e0af5bf3ddab5b67508003ca47678cd

### Kernel Version

none

### Kubernetes Version

none

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

I think need add new method `UpdateInstance(ctx context.Context, obj *v2.CiliumNode) error` for `instanceManager` that call cloud api to get instance for this node and update `instanceManager.instances` .  do update instances logic in `n.ops.UpdatedNode(resource)`
https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/ipam/node.go#L355-L377

``` go
// UpdatedNode is called when an update to the CiliumNode is received.
func (n *Node) UpdatedNode(obj *v2.CiliumNode) {
	n.mutex.Lock()
	n.k8sObj = obj
        n.mutex.Unlock()
        err := n.manager.UpdateInstance(context.Todo(), obj)
        ..........
}
```
other ideas is  add new method  `UpdateInstance(ctx context.Context, obj *v2.CiliumNode) error`   for  `AllocationImplementation` interface and do logic `node.UpdatedResource`
``` go
func (n *Node) UpdatedResource(resource *v2.CiliumNode) bool {
	// Deep copy the resource before storing it. This way we are not
	// dependent on caller not using the resource after this call.
	resource = resource.DeepCopy()

	n.ops.UpdatedNode(resource)
        err := n.manager.instancesAPI.UpdateInstance(context.Todo(), obj)
        if err != nil {
           return false
         }

	n.mutex.Lock()
	// Any modification to the custom resource is seen as a sign that the
	// instance is alive
	n.instanceRunning = true
	n.resource = resource
	n.mutex.Unlock()

	n.recalculate()
	allocationNeeded := n.allocationNeeded()
	if allocationNeeded {
		n.requirePoolMaintenance()
		n.poolMaintainer.Trigger()
	}

	return allocationNeeded
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by wu0407**
Update the instance network interface information when a new node is created.
This is fix errors return by `ResyncInterfacesAndIPs` "Instance not found! Please delete
corresponding ciliumnode if instance has already been deleted" when new node
created in cloud provider then add to cluster and before resync triggered.

Fixes: #20678

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

```release-note
Eliminate a delay between new node creation and creation of ENIs
```

### CODE REVIEW

File: pkg/ipam/types/types.go [hunk 1]
```diff
@@ -482,3 +482,20 @@ func (m *InstanceMap) NumInstances() (size int) {
 	m.mutex.RUnlock()
 	return
 }
+
+// IsExist returns whether the instance ID is in the instanceMap
+func (m *InstanceMap) IsExist(instanceID string) (exist bool) {
```
Comment (gandro) [inline pkg/ipam/types/types.go]: nit:

```suggestion
// Exists returns whether the instance ID is in the instanceMap
func (m *InstanceMap) Exists(instanceID string) (exists bool) {
```

File: pkg/ipam/node_manager.go [hunk 2]
```diff
@@ -115,6 +115,12 @@ type AllocationImplementation interface {
 	// chance to resync its own state with external APIs or systems. It is
 	// also called when the IPAM layer detects that state got out of sync.
 	Resync(ctx context.Context) time.Time
+
+	// UpdateInstance is sync interfaces info for new instance in instances if new instance not exist
+	UpdateInstance(ctx context.Context, obj *v2.CiliumNode) time.Time
+
+	// DeleteInstance is delete instance form instances
```
Comment (gandro) [inline pkg/ipam/node_manager.go]:
```suggestion
	// DeleteInstance deletes the instance from instances
```

**Comment by gandro** [comment]
Thanks for the PR. Unfortunately I don't fully understand the problem this PR is trying to address: The `Instance not found! Please delete
corresponding ciliumnode if instance has already been deleted` error "fixed" here is temporary, right? Eventually it will fix itself when a resync happens and this PR "only" tries to shorten that timeframe.

Unfortunately I find this solution very hard to follow. Would it be possible to force the resync to create missing instance in `InstancesManager.CreateNode` instead? That way we always know that any existing node has an instance behind it?

**Comment by wu0407** [comment]
> Thanks for the PR. Unfortunately I don't fully understand the problem this PR is trying to address: The `Instance not found! Please delete corresponding ciliumnode if instance has already been deleted` error "fixed" here is temporary, right? Eventually it will fix itself when a resync happens and this PR "only" tries to shorten that timeframe.

yes, that fix if new node created in cloud provider then add to cluster and before resync triggered.
It not fixed  `Instance not found! Please delete corresponding ciliumnode if instance has already been deleted` error because instance and subnet do not match Filter rules.

> 
> Unfortunately I find this solution very hard to follow. Would it be possible to force the resync to create missing instance in `InstancesManager.CreateNode` instead? That way we always know that any existing node has an instance behind it?

Put `updateInstance` in `InstancesManager.CreateNode` better than `node.UpdatedResource`, I will change to do this. But it also need to check instance whether exist in instances. such as old instance in aws ec2 list, but not in cluster that aleady in `InstancesManager.instances`, so it add to cluster, no need to resync instances.

**Comment by wu0407** [comment]
I re-implementation check instance exist in CreateNode

File: pkg/ipam/node_manager.go [hunk 3]
```diff
@@ -279,7 +282,13 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
 			logLimiter:          logging.NewLimiter(10*time.Second, 3), // 1 log / 10 secs, burst of 3
 		}
 
-		node.ops = n.instancesAPI.CreateNode(resource, node)
+		var instanceExist bool
+		node.ops, instanceExist = n.instancesAPI.CreateNode(resource, node)
+		if !instanceExist {
+			if _, ok := n.instancesAPIResync(context.TODO()); !ok {
+				node.logger().Warning("Failed to resync instance API after found new node")
```
Comment (gandro) [inline pkg/ipam/node_manager.go]:
```suggestion
				node.logger().Warning("Failed to resync instance API after new node was found")
```

File: pkg/ipam/node_manager.go [hunk 4]
```diff
@@ -347,7 +356,13 @@ func (n *NodeManager) Delete(nodeName string) {
 		}
 	}
 
-	delete(n.nodes, nodeName)
+	// Delete instance from instanceManager in case Node.recalculate() use stale data
+	// when nodes from being deleted and then rejoined into the cluster
```
Comment (gandro) [inline pkg/ipam/node_manager.go]: Can you elaborate on that comment? I don't see how removing the node from the instances list affects `n.recalcuate()`. Can you point me to the code which will behave differently? Thanks!

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the updated version. I didn't realize that we can't call `instancesAPIResync` from the `InstancesManager`. So maybe keeping `CreateNode` is it was previously might be cleaner.

I think we're getting close though. Looking at the code again, I feel the the following might not be terrible:

First, let's undo the change to `CreateNode` again (sorry for that). Instead, let's change the `AllocationImplementation` to have a `HasInstance` and `DeleteInstance` method. The `DeleteInstance` is already there, the `HasInstance` checks if the instance exists and returns a boolean.

In `NodeManager.Update`, let's then trigger the resync if the detect that an instance is missing, for example:

```golang
if !n.instancesAPI.HasInstance(resource.InstanceID()) {
	if _, ok := n.instancesAPIResync(context.TODO()); !ok {
		node.logger().Warning("Failed to resync instance API after new node was found")
	}
}

node.ops = n.instancesAPI.CreateNode(resource, node)
```

I also added some inline comments regarding language.

File: pkg/ipam/node_manager.go (hunk 4 above)
Comment (wu0407) [reply pkg/ipam/node_manager.go]: Such as I want downgrade EC2 instance configuration，so remove node from cluster and change instance type, then rejoined in the cluster. The number of instance bound ENIs may exceed the limit, it need unbound ENI. The instance of enis is changed. If the node rejoins the cluster before next `instanceAPIResync`,  `ResyncInterfacesAndIPs` return incorrect result.

Comment (gandro) [reply pkg/ipam/node_manager.go]: Ah, so this is relying on the code added by this PR (which forces an immediate resync if the instance is not in the list). Makes sense. 

Let's word it a little bit differently though, because this wasn't clear to me just from the comment:

```suggestion
	// Delete the instance from instanceManager. This will cause Update() to
	// invoke instancesAPIResync if this instance rejoins the cluster.
	// This ensures that Node.recalculate() does not use stale data for
	// instances which rejoin the cluster after their EC2 configuration has changed.
```

**Review by gandro — COMMENTED** [review]
Looking good! One last suggestion here: https://github.com/cilium/cilium/pull/21027#discussion_r957231113

**Comment by wu0407** [comment]
@gandro  done

**Review by gandro — APPROVED** [review]
Thank you!

**Review by joestringer — APPROVED** [review]
Approving on behalf of @cilium/alibabacloud

File: pkg/ipam/node_manager.go [hunk 5]
```diff
@@ -279,6 +285,12 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
 			logLimiter:          logging.NewLimiter(10*time.Second, 3), // 1 log / 10 secs, burst of 3
 		}
 
+		if !n.instancesAPI.HasInstance(resource.InstanceID()) {
+			if _, ok := n.instancesAPIResync(context.TODO()); !ok {
+				node.logger().Warning("Failed to resync instance API after new node was found")
```
Comment (christarazi) [inline pkg/ipam/node_manager.go]:
```suggestion
				node.logger().Warning("Failed to resync the instances from the API after new node was found")
```

**Review by christarazi — COMMENTED** [review]
Thanks for the fix! Could you update the commit msg to be more descriptive on what the fix is doing and why?

File: pkg/ipam/node_manager.go (hunk 5 above)
Comment (wu0407) [reply pkg/ipam/node_manager.go]: > Thanks for the fix! Could you update the commit msg to be more descriptive on what the fix is doing and why?

The updated commit message
```
eni: fix new node not immediately triggers the creation of ENI

1. Sync the instances network interface information from instances api when a new node is created
but not in `instanceManager.instances`. This is fix errors in `ResyncInterfacesAndIPs` "Instance not found! Please delete
corresponding ciliumnode if instance has already been deleted" when new node
created in cloud provider then add to cluster and before resync triggered.

2. Delete instance from `instanceManager.instances` on node deleted. This will cause `NodeManager.Update()` to
invoke `instancesAPIResync` if this instance rejoins the cluster.
This ensures that `Node.recalculate()` does not use stale data for
instances which rejoin the cluster after their EC2 configuration has changed.

Fixes: #20678
```

**Comment by wu0407** [comment]
> Thanks for the fix! Could you update the commit msg to be more descriptive on what the fix is doing and why?

The failed ci test  like Flake  that need retest. @christarazi

**Review by christarazi — APPROVED** [review]

**Comment by christarazi** [comment]
Marking ready to merge because we have approving reviews and CI has passed already.
